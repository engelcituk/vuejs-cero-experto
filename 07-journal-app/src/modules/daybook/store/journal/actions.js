
import journalApi from '@/api/journalApi'
/*
export const myAction = async ({commit}) => {

}
*/

export const loadEntries = async ({commit}) => {
   const { data } = await journalApi.get(`/entries.json`)
   const entries = []
   for(let id of Object.keys(data)){
       entries.push({
           id,
           ...data[id]
       })
   }
   commit('setEntries', entries)
}

export const updateEntry = async ({commit}, entry) => {
    const { id, date, picture, text } = entry
    const dataToSave = { date, picture, text }
    commit('setLoading', true)
    await journalApi.put(`/entries/${id}.json`, dataToSave)
    commit('updateEntry', {...entry})
    commit('setLoading', false)
}

export const createEntry = async ({commit}, entry) => {
    const { date, picture, text } = entry
    const dataToSave = { date, picture, text }
    commit('setLoading', true)
    const { data } = await journalApi.post(`/entries.json`, dataToSave)
    dataToSave.id = data.name
    console.log(dataToSave)
    commit('addEntry', dataToSave)
    commit('setLoading', false)
    return data.name
}