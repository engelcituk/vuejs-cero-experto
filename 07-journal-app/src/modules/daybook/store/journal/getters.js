/*
    export const myGetter =  (state) => {
        return state.something
    }
*/

export const getEntriesByTerm =  (/*state*/) => {
    // return state.something
}

export const getEntryById =  (/*state*/) => {
    // return state.something
}