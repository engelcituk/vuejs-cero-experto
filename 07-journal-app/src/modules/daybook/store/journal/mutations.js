/*
    export const myMuttation =  (state) => {
        return state.something
    }
*/

export const setLoading =  (state, bool) => {
    state.isLoading = bool
}

export const setEntries =  (state, entries) => {
  state.entries = [...state.entries, ...entries] // crear un arreglo con los valoes antiguos y los nuevos entries
  state.isLoading = false
}

export const updateEntry =  (state, entry) => {
    const idx = state.entries.map( e => e.id ).indexOf(entry.id)
    state.entries[idx] = entry
}

export const addEntry =  (/*state*/) => {
    // return state.something
}