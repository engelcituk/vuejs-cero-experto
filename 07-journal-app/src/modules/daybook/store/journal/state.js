export default () => ({
    isLoading: true,
    entries:[
        {
            id: '1',
            date: new Date().toDateString(),
            text:'Lorem ipsum dolor sit amet consectetur adipisicing elit. Quisquam quaerat ex quasi quam? Assumenda voluptas tenetur tempore incidunt ad consequatur delectus mollitia veritatis ducimus placeat rerum, dignissimos deleniti, officiis cumque!',
            picture: null
        },
        {
            id: '2' ,
            date: new Date().toDateString(),
            text:'Tempor quis excepteur esse elit aliquip et ad anim aliquip veniam consequat nostrud elit qui. Consectetur deserunt do sint incididunt exercitation duis. Fugiat adipisicing in sint occaecat esse. Ea nulla ad proident nulla voluptate elit amet aliquip reprehenderit occaecat id ullamco. Id et magna aute sunt ut anim pariatur culpa commodo est voluptate sint.',
            picture: null
        },
        {
            id: '3' ,
            date: new Date().toDateString(),
            text:'Ex dolore nostrud exercitation occaecat pariatur. Irure esse tempor nulla mollit Lorem quis consectetur ea laboris dolor nisi id sunt aute. Duis tempor ex exercitation enim nulla ea reprehenderit id cupidatat esse.',
            picture: null
        }
    ]
})