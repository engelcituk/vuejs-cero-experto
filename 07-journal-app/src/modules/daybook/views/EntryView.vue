<template>
    <template v-if="entry">
        <div
            class="entry-title d-flex justify-content-between p-2">
            <div>
                <span class="text-success fs-3 fw-bold">{{day}}</span>
                <span class="mx-1 fs-3">{{month}}</span>
                <span class="mx-2 fs-4 fw-light">{{yearDay}}</span>
            </div>
            <div>
                <input
                    v-show="false"
                    type="file"
                    @change="onSelectedImage"
                    ref="imageSelector"
                    accept="image/png, image/jpeg"
                >
                <button class="btn btn-danger mx-2" @click="onDeleteEntry" v-if="entry.id">
                    Borrar
                    <i class="fa fa-trash-alt"></i>
                </button>
                <button class="btn btn-primary " @click="onSelectImage">
                    Subir foto 
                    <i class="fa fa-upload"></i>
                </button>
            </div>
        </div>
        <hr>
        <div class="d-flex flex-column px-3 h-75">
            <textarea
                v-model="entry.text"
                placeholder="¿Que sucedió hoy?"
            ></textarea>
        </div>
        
        <!-- <img
            class="img-thumbnail"
            width="500"
            height="450"
            src="http://t0.gstatic.com/licensed-image?q=tbn:ANd9GcTXTItGe_WhSapgSydJU92LCgSWszJ4j_JkMuU5revTBQAGtF4KI4za1rR018f0s0mCfCUxGvWfaADHV8wyE_c" alt="entry pictture"
        > -->

        <img
            v-if="localImage"
            :src="localImage"
            class="img-thumbnail"
            width="500"
            height="450"
            alt="entry pictture"
        >
        <Fab
            :icon="'fa-save'"
            @save-entry="saveEntry"
        />
    </template>
</template>

<script>
    import Swal from 'sweetalert2'
    import { defineAsyncComponent } from 'vue';
    import { mapGetters, mapActions } from 'vuex'
    import getDayMonthYear from './../helpers/getDayMonthYear'

    export default {
        props:{
            id:{
                type: String,
                required: true
            }
        },
        components:{
            Fab: defineAsyncComponent( ()=> import('./../components/Fab')) //lazy load component
        },
        data(){
            return {
                entry: null,
                localImage: null,
                file: null,
            }
        },
        created(){
            this.loadEntry()
        },
        computed:{
            ...mapGetters('journal',['getEntryById']),
            day(){
                const { day } = getDayMonthYear(this.entry.date)
                return day
            },
            month(){
                const { month } = getDayMonthYear(this.entry.date)
                return month
            },
            yearDay(){
                const { yearDay } = getDayMonthYear(this.entry.date)
                return yearDay
            }
        },
        methods:{
            ...mapActions('journal',['createEntry', 'updateEntry','deleteEntry']),
            loadEntry(){
                let entry
                if( this.id === 'new' ){
                    entry = {
                        text: '',
                        date: new Date().getTime()
                    }

                } else {
                     entry = this.getEntryById(this.id)
                    if( !entry ) return this.$router.push({name:'no-entry'})
                }
                this.entry = entry
            },
            async saveEntry(){
                new Swal({
                    title: 'Espere por favor',
                    allowOutsideClick: false,
                })
                Swal.showLoading()

                if( this.entry.id ){
                    //actualizar
                    await this.updateEntry(this.entry)
                } else {
                    const id = await this.createEntry(this.entry)
                    this.$router.push({name:'entry', params:{ id }})
                }
                Swal.fire('Guardado', 'Entrada registrada con éxito', 'success')
            },
            async onDeleteEntry(){
                const { isConfirmed } = await Swal.fire({
                    title: '¿Está seguro?',
                    text:'Una vez borrado no podrá recuperarlo',
                    showDenyButton: true,
                    confirmButtonText: 'Sí borrar',
                    denyButtonText: 'Cancelar',
                })
                if( isConfirmed ){
                    new Swal({
                        title: 'Espere por favor',
                        allowOutsideClick: false,
                    })
                    Swal.showLoading()

                    await this.deleteEntry(this.entry.id)
                    this.$router.push({name:'no-entry'})
                    Swal.fire('Borrado', '', 'success')
                }
            },
            onSelectedImage(event){
                const file = event.target.files[0]
                if( !file ){
                    this.localImage = null
                    this.file = null
                    return
                }
                this.file = file
                const fr = new FileReader()
                fr.onload = () => this.localImage = fr.result
                fr.readAsDataURL( file )
            },
            onSelectImage(){
                this.$refs.imageSelector.click()
            }
        },
        watch:{
            id(){
                this.loadEntry()
            }   
        }
    }
</script>

<style lang="scss" scoped>
    textarea{
        font-size: 20px;
        border: none;
        height: 100%;

        &:focus{
            outline: none;
        }
    }
    img{
        width: 200px;
        position: fixed;
        bottom: 150px;
        right: 20px;
        box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2)
    }
</style>