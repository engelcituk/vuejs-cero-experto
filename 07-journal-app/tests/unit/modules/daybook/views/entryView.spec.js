
import { shallowMount } from '@vue/test-utils'
import { createStore } from 'vuex'

import journal from '@/modules/daybook/store/journal'
import EntryView from '@/modules/daybook/views/EntryView'
import { journalState } from './../../../mock-data/test-journal-state'

const createVuexStore = (initialState) => 
    createStore({
        modules:{
            journal:{
                ...journal,
                state: { ...initialState }
            }
        }
    })
    
describe('Pruebas en el EntryView', () => {

    const store = createVuexStore(journalState)
    const mockRouter = {
        push: jest.fn()
    }


    let wrapper

    beforeEach( () => {
        jest.clearAllMocks()
        wrapper = shallowMount(EntryView, {
            props:{
                id: '-MjIFKf-5JWdJFgyqNXA',

            },
            global:{
                mocks:{
                    $router: mockRouter
                },
                plugins:[ store ]
            }
        })
    })

    test('Debe de sacar al usuario porque el id no existe', ()=> {
       const wrapper = shallowMount(EntryView, {
           props:{
               id: 'Este id no existe en el store'
           },
            global:{
                mocks:{
                    $router: mockRouter
                },
                plugins:[ store ]
            }
        })

        expect( mockRouter.push ).toHaveBeenCalledWith({name:'no-entry'})
    })   

    test('Debe de mostrar la entrada correctamente', ()=> {
        expect( wrapper.html() ).toMatchSnapshot()
        expect( mockRouter.push ).not.toHaveBeenCalled()
    })
      
})