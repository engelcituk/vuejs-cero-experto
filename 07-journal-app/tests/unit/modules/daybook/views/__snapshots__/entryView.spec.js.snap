// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el EntryView Debe de mostrar la entrada correctamente 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">11</span><span class="mx-1 fs-3">Septiembre</span><span class="mx-2 fs-4 fw-light">2021, Sábado</span></div>
  <div><input type="file" accept="image/png, image/jpeg" style="display: none;"><button class="btn btn-danger mx-2"> Borrar <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary "> Subir foto <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="¿Que sucedió hoy?"></textarea></div><img src="https://res.cloudinary.com/di3qzyeke/image/upload/v1631340833/q0yw6icrapvwcbw8kgxx.jpg" class="img-thumbnail" width="500" height="450" alt="entry pictture">
<!--v-if-->
<fab-stub icon="fa-save"></fab-stub>
`;
